util/pyclient/bft_client.py

```diff
@@ -241,3 +247,256 @@ def get_rsi_replies(self):
+        """
+        server_cert_path = self._get_cert_path(dest_replica.id, False)
+        client_cert_path = self._get_cert_path(self.client_id, True)
+        client_pk_path = self._get_private_key_path(self.client_id, True)
+        # Create an SSl context - enable CERT_REQUIRED and check_hostname = True
+        ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
+        # Verify server certificate using this trusted path
+        ssl_context.load_verify_locations(cafile=server_cert_path)
+        # Load my private key and certificate
+        ssl_context.load_cert_chain(client_cert_path, client_pk_path)
+        # Server hostname to be verified must be taken from create_tls_certs.sh
+        server_hostname = self.CERT_DOMAIN_FORMAT % dest_replica.id
+        while True:
+            try:
+                # Open TCP stream and connect to server
+                tcp_stream = await trio.open_tcp_stream(str(dest_replica.ip), int(dest_replica.port))
+                # Wrap this stream with SSL stream, pass server_hostname to be verified
+                ssl_stream = trio.SSLStream(tcp_stream, ssl_context, server_hostname=server_hostname, https_compatible=False)
+                # Wait for handshake to finish (we want to be on the safe side - after this we are sure 
+                # connection is open)
+                await ssl_stream.do_handshake()
+                # Success! keep stream in dictionary and break out
+                self.ssl_streams[(dest_replica.ip, dest_replica.port)] = ssl_stream
+                return
+            except (OSError, trio.BrokenResourceError):
+                await trio.sleep(0.25)
+                pass
+            continue
+
+    async def on_stopped_replicas(self, replica_ids):
+        """ BFT network inform client when a group of replicas are stopped. We relay on caller cancel scope """
+        for id in replica_ids:
+            replica = self.replicas[id]
+            addr = (replica.ip, replica.port)
+            if addr in self.ssl_streams:
+                if addr in self.ssl_streams:
+                    stream = self.ssl_streams[addr]
+                    del self.ssl_streams[addr]
+                    await stream.aclose()
+
+    async def _send_data(self, data, dest_replica):
+        """ Try to send data to dest_replica. On exception - close the SSL stream. """
+        dest_addr = (dest_replica.ip, dest_replica.port)
+        if dest_addr not in self.ssl_streams.keys():
+            # dest_replica is not connected (crushed? isolated?) 
+            return
+        # first 4 bytes include the data header (message size), then comes the data
+        data_len = len(data)
+        out_buff = bytearray(data_len.to_bytes(self.MSG_HEADER_SIZE, "little"))
+        out_buff += bytearray(data)
+        stream = self.ssl_streams[dest_addr]
+        try:
+            await stream.send_all(out_buff)
```

Comment:
don't we need a timeout here?

Reply:
No.
all the receive and send are covered by large trio scope:
with trio.move_on_after(self.config.retry_timeout_milli / 1000):
this one do retries.